### PR TITLE
Advocate for functions as parameters

### PR DESCRIPTION
**Pull request by lgvalle**
At the moment, our style guides for Kotlin discourages using functions as parameters for classes.

I want to challenge that recommendation with this PR. The reasons are described in the modified markdown document, so I won't repeat them here. Please read them [there](https://github.com/novoda/novoda/pull/61/files).

The aim is to have a productive discussion here about this matter and either adopt the change or discard it, but in any case, the reasoning will stay in this PR and could be used as future reference.

### CODE REVIEW

File: style-guides/kotlin/README.md [hunk 1]
```diff
@@ -240,13 +240,38 @@ This depends on its visibility. If we have a `private` sealed class then it is s
...
+    durationCalculator: (Int) -> Int) {}
+```
+
+These funtions are difficult to work with: we don't know what they're doing from the signature and they are prone to error because they entry and return types are not restricted to a domain. 
+
+### Good example of parameter functions:
+```kotlin
+class GlobalVariablesProducer(
+    videoIdParser: (RawResponseId) -> VideoId,
+    ageRater: (UserAge) -> AgeRating,
```
Comment (zegnus) [inline style-guides/kotlin/README.md]: This approach (in class constructors) to me has four main flaws that make maintenance a nightmare:
1. There is no way to know how many different providers of this function we have
2. If there is a bug you cannot directly go to who implements that because of its a function.
3. If we need to modify the params, for example now I want to pass two parameters, you need to change this signature. This breaks SRP.
4. The implementation is done raw in its factory (the one that provides the function) and factories should just provide dependencies, not implementation details.

Comment (pablisco) [reply style-guides/kotlin/README.md]: Some thoughts on this :D 

1. This usage of interfaces can actually cause issues. I know of a case of a simple (SAM) interface, let's call it `Feature`, which has dozens of implementations. Instead, in many cases, we can use sealed classes to describe all the cases available and their required data. then we have a single implementation that uses that data. Makes the code easier to follow IMHO.
2. If this relates to the Error report, we still get a stack trace of where the error happened. Also, if we put a breakpoint where the function is called we can jump inside the function and the IDE will send us to where the function is implemented.
3. Not sure where the relation to the Single Responsibility Principle is. If we add a new parameter to the method in the type we also need to change the usage of the method wherever it is used. If we use the IDE there is an option to change the type and cascade changes. A counterpoint to this is that if we want to change the signature only in one place, we have the flexibility to do so without affecting other locations or adding a new type.
4. We don't need as many factories, the function can be defined in the constructor as a default value which means that we can change it in testing. And if we want to provide a different implementation, we can do so without changing the class itself. Open for extension, closed to modification.

File: style-guides/kotlin/README.md [hunk 2]
```diff
@@ -240,13 +240,38 @@ This depends on its visibility. If we have a `private` sealed class then it is s
...
+  - It simplifies testing: a class receiving a function as a parameter knows for sure that this collaborator is stateles. It's very easy to provide a fake function for testing purposes. 
+
+**NOPE**: Functions that transform from basic types, such as `(Int) -> String` or `(String, Int) -> Boolean` are **NOT** good entry parameter types. In those cases, it's very hard to identify what this function does or where's the implementation. 
+### Bad example of parameter functions:
+
+```kotlin
+class GlobalVariablesProducer(
+    videoIdParser: (Int) -> String,
+    ageRater: (Int) -> Boolean,
+    durationCalculator: (Int) -> Int) {}
```
Comment (pablisco) [inline style-guides/kotlin/README.md]: This example code and the next, let's have the `) {}` as the parent style guides dictates: https://developer.android.com/kotlin/style-guide#functions

File: style-guides/kotlin/README.md [hunk 3]
```diff
@@ -240,13 +240,38 @@ This depends on its visibility. If we have a `private` sealed class then it is s
...
+These funtions are difficult to work with: we don't know what they're doing from the signature and they are prone to error because they entry and return types are not restricted to a domain. 
+
+### Good example of parameter functions:
+```kotlin
+class GlobalVariablesProducer(
+    videoIdParser: (RawResponseId) -> VideoId,
+    ageRater: (UserAge) -> AgeRating,
+    durationCalculator: (TimeInMillis) -> VideoDuration) {}
+```
+These functions are easy to work with: we know exaclty what they are doing by just reading the signature. Their types are strong and bound to our domain, so it's harder to pass a wrong value to them.
```
Comment (pablisco) [inline style-guides/kotlin/README.md]: I would add that we should aim to use data classes that wrap primitive values to give them semantics:
```kotlin
data class UserAge(val age: Int)
```
In the future, when `inline classes` are out of experimental we should use them instead. May be worth creating a dummy annotation in the mean time to it's easier to replace them in the future.
```kotlin
@FutureInline
data class UserAge(val age: Int)
```

File: style-guides/kotlin/README.md (hunk 1 above)
Comment (lgvalle) [reply style-guides/kotlin/README.md]: Thanks for offering another point of view @zegnus. Those four points are correct. But let's see if I can convince you they don't necessarily affect us:

> 1. There is no way to know how many different providers of this function we have

Why do we care really? From the point of view of this class, it's like receiving an interface: as long as it sticks to the contract we should not care here who implements this. 
If we need to find out what's the actual implementation received here we can always navigate to wherever this constructor is invoked. That's the exact same thing we'll do with an interface. If there are multiple classes implementing an interface we need to go to where this constructor is called to find out what's the specific one injected.

> 2. If there is a bug you cannot directly go to who implements that because of its a function.

I think this falls into the same category as the above. Eventually, you need to navigate to the class constructor invocation to find out what's been injected here. 

> 3. If we need to modify the params, for example now I want to pass two parameters, you need to change this signature. This breaks SRP.

I don't see why this breaks SRP:

_Every module, class, or function should have responsibility over a single part of the functionality provided by the software, and that responsibility should be entirely encapsulated by the class._

Uncle Bob famous way to express this was: _"A class should have only one reason to change"_

And here I can see why you think it breaks SRP. What I'd argue is this: if you need to modify the signature because now the collaborator function needs a new parameter, then that's a sign your types are not strong enough, and because of that, your classes are not decoupled enough. 

Let's say in the example above we find that we need to change this collaborator function:
```kotlin
videoIdParser: (RawResponseId) -> VideoId
``` 
because now, to generate a `VideoId` we also need, lets say, the video url. So we need something like:
```kotlin
videoIdParser: (RawResponseId, URL) -> VideoId
```

The problem here is on the types, and because of that, your `GlobalVariablesProducer` knows too much about its collaborators. If we refactor this to something like this:

```kotlin
videoIdParser: (RawVideoData) -> VideoId
```
we can include any data we need for the actual parsing inside that data class. And future changes won't affect the class using this function.

TL;DR: a collaborator signature change is a smell of not strong enough types IMHO

> 4. The implementation is done raw in its factory (the one that provides the function) and factories should just provide dependencies, not implementation details.

You can do it that way, but you can also have a file with top-level functions implementing that signature. You can, in fact, have multiple functions implementing the same signature in the same file, if you need multiple implementations. And this makes points 1 and 2 mentioned above easier with functions than with interfaces+classes because everything is in the same file

Comment (ouchadam) [reply style-guides/kotlin/README.md]: also got some thoughts... 😛 

- the function instances are unique per build which means they don't stack correctly across builds in `crashlytics` as they end up as `OuterClass.lambda$QPm1yLbKIuayUbRUJJ(unique per build)`

- Decoupling is helpful for cross module communication where there's no common parent

- SRP breaking can be avoided by currying 
```kotlin
// (URL) -> ( (Response) -> VideoId )

val providerFactory = { url -> 
 { response -> VideoId(url, response) }
}
val provider = providerFactory(url)
```

File: style-guides/kotlin/README.md (hunk 3 above)
Comment (amlcurran) [reply style-guides/kotlin/README.md]: +100

**Comment by ataulm** [comment]
I was just reviewing one of my PRs I noticed I did something that looks weird.

I have a class:

```kotlin
class FeedAdapter(private val onClickHeader: () -> Unit): ListAdapter...
```

and at the call site, I did this:

```kotlin
val adapter = FeedAdapter { analyticsService.track("blah blah") }
```

but it's not clear what's happening. I think AS will recommend doing the above but I think it's clearer (for class constructors) to say it explicitly:

```kotlin
val adapter = FeedAdapter(onClickHeader = { analyticsService.track("blah blah") })
```
or

```kotlin
val onClickHeader = { analyticsService.track("blah blah") }
val adapter = FeedAdapter(onClickHeader)
```

It might be worth adding guidelines around this (one way or the other) in this PR too.

File: style-guides/kotlin/README.md (hunk 1 above)
Comment (zegnus) [reply style-guides/kotlin/README.md]: @lgvalle and @pablisco you are implying that my alternative for a function as a param is an interface and I didn't say that.

I see and appreciate the value on functions as params and I understand the idea that's behind it, mainly a functional programming approach with state-less code. My beef is not on that, the problem that I have in functions as **params in class constructors** is that it makes it more complicated to follow the code and to find out what's happening when you need to solve bugs and follow calls from one class to another, on top of what @ouchadam also has mentioned. And I have the same beef with overusing interfaces as I expose here https://github.com/novoda/novoda/blob/master/style-guides/kotlin/README.md#interfaces, so this is not about interfaces vs functions, is about making the life of the developer easier.

Passing a function as a param makes it impossible to click on it in order to follow the implementation that's behind it. When you click on the usage of that function, the further that you are going to go is to the constructor. This to me is a major drawback and if it's not for any reader of this post, it means that you haven't faced this problem with multiple classes calling multiple functions as params and going crazy trying to find out what's actually being implemented.

In terms of who is providing the implementation, to me that's also a sticky point. I am not sure that the constructor should directly provide a function implemented in there, or in the class itself. All dependencies should be injected, it doesn't matter if its a function or not.

Functions as params in methods with one param is a different thing because you can actually click on who calls that method and you can see its implementation easily. I'm pointing this out in case you think I'm against functional programming, I'm not xD

This major drawback should be compensated by a major advantage, and unfortunately, I don't see. While I don't see it I prefer to keep it simple = pass a class and think twice if you really need an interface in there (can you pass a function? yes you can, should you? no I don't think so)

Comment (pablisco) [reply style-guides/kotlin/README.md]: Yeah, it is true that it can be hard if we just have a lambda out of the blue. This is a problem only if we use Factories to create the object. In Kotlin we can do the following though:
```kotlin
class GlobalVariablesProducer(
    videoIdParser: (RawResponseId) -> VideoId = videoIdParser(),
    ageRater: (UserAge) -> AgeRating = defaultAgeRater,
    durationCalculator: (TimeInMillis) -> VideoDuration = durationCalculator
) {}

// somewhere else (or there as private)
val defaultAgeRater: (UserAge) -> AgeRating = { ... }
val durationCalculator: (TimeInMillis) -> VideoDuration = { ... }
fun videoIdParser(
  videoCache: () -> Video = videoCache
): (TimeInMillis) -> VideoDuration = { ... }
val videoCache: () -> Video = { ... }
// These are stateless and can live as top level functions. If they require other operations they can also be defined as
```

With factories and interfaces we can click to the implementation of the interface. The moment we have more than one implementation, then we have to navigate to the factory to see what implementation is used. In which case we have to either know where the factory is or look for usages of `GlobalVariablesProducer` and look for the place where the component is created. It can be used on several places as well as in tests. It can get a bit of a long process, specially when having multiple nesting of factories within factories. 

When dealing with factories, more than once I've been lost in a trail of factories looking for an implementation. To the point that I've forgotten why I was after that implementation 😅 

Using default values instead, as suggested:
- We can navigate to the dependency very easily and directly
- In testing we can replace the dependencies in the same way or even easier

Maybe we should add this recommendation as a way to solve the navigation issues as well as a way to reduce boilerplate, cognitive load and navigation trails.

The only drawback that I can see is that anyone can initiate the class with other implementations of the given function. Specially since there is no package protected visibility in Kotlin. Although, arguably that could encourage to have smaller modules (instead of packages) and use `internal` visibility instead.

Does this look like a good alternative?

Extra: Using private functions on the same file also allows us to give a component an extra level of encapsulation without having to expose classes just for testing. As we would only need to expose the function.

**Comment by pablisco** [comment]
@ataulm Nice point. I've observed that too. I wish there was a way to force names for lambdas when we don't want them to be blocks. But I guess we need to have them on guidelines for now 😅  (Maybe a compiler plugin? 🤔 )

Only thing that I would add is that if we have a function with a single method from Java, the problem is the same:

```java
// in java
interface Clicker {
  void click()
}
```
```kotlin
// in kotlin
class FeedAdapter(private val onClickHeader: Clicker): ListAdapter...
```

It would let us do:
```kotlin
val adapter = FeedAdapter { analyticsService.track("blah blah") }
```

Also, it is not the same if the interface is defined in Kotlin as it forces us to do it like this:

```kotlin
val adapter = FeedAdapter(object : Clicker {
    override fun click() {
        analyticsService.track("blah blah")
    }
})
```

Which is actually worse 😅

💯 to add some details about how to deal with these. I wish the IDE would suggest adding a name as well as extracting the lambda :D

**Comment by charroch** [comment]
My rusty 0.02$ that might or might not help in the thinking:

- A class should encapsulate logic and data together.
- A class with no data (i.e. state) is probably not a class
- The uniqueness of a class (i.e instance) should depend on its state not behaviour 
- A good practice is to overwrite the ```equal``` and ```hashCode``` methods to surface the above
- Anything that does not participate in the ```equal``` or ```hashCode``` should be injected as dependencies (or extracted out of the class)

File: style-guides/kotlin/README.md [hunk 4]
```diff
@@ -240,13 +240,38 @@ This depends on its visibility. If we have a `private` sealed class then it is s
...
+```
+
+These funtions are difficult to work with: we don't know what they're doing from the signature and they are prone to error because they entry and return types are not restricted to a domain. 
+
+### Good example of parameter functions:
+```kotlin
+class GlobalVariablesProducer(
+    videoIdParser: (RawResponseId) -> VideoId,
+    ageRater: (UserAge) -> AgeRating,
+    durationCalculator: (TimeInMillis) -> VideoDuration) {}
```
Comment (pablisco) [inline style-guides/kotlin/README.md]:
```suggestion
    durationCalculator: (TimeInMillis) -> VideoDuration
) {}
```

**Comment by lgvalle** [comment]
This PR is probably not needed anymore folks ❤️

**Comment by rock3r** [comment]
Lgtm

**Comment by friedger** [comment]
❤️

**Comment by blundell** [comment]
fwiw I would have shipped this 😆
